examples/email-assistant/application.py

```diff
@@ -0,0 +1,174 @@
+        "feedback",
+    ],
+    writes=["current_draft", "draft_history"],
+)
+def formulate_draft(state: State) -> Tuple[dict, State]:
+    """Formulates the draft response based on the incoming email, response instructions, and any clarifications."""
+    # TODO -- query an LLM to generate the draft response
+    incoming_email = state["incoming_email"]
+    response_instructions = state["response_instructions"]
+    client = _get_openai_client()
+    # TODO -- use instructor to get a pydantic model
+    clarification_answers_formatted_q_a = "\n".join(
+        [f"Q: {q}\nA: {a}" for q, a in state["clarification_answers"].items()]
+    )
+    instructions = [
+        f"The email you are to respond to is: {incoming_email}.",
+        f"Your instructions are: {response_instructions}.",
+        "You have already asked the following questions and received the following answers: ",
+        clarification_answers_formatted_q_a,
+    ]
+    if state["draft_history"]:
+        instructions.append("Your previous draft was: ")
+        instructions.append(state["draft_history"][-1])
+        instructions.append(
+            "you received the following feedback, please incorporate this into your response: "
+        )
+        instructions.append(state["feedback"])
+    instructions.append("Please generate a draft response using all this information!")
+    prompt = " ".join(instructions)
+
+    result = client.chat.completions.create(
+        model="gpt-4",
+        messages=[
+            {
+                "role": "system",
+                "content": "You are a chatbot that has the task of generating responses to an email. ",
+            },
+            {"role": "user", "content": prompt},
+        ],
+    )
+    content = result.choices[0].message.content
+    return {"prompt": prompt, "current_draft": content}, state.update(current_draft=content).append(
+        draft_history=content
+    )
+
+
+@action(reads=[], writes=["feedback"])
+def process_feedback(state: State, feedback: str) -> Tuple[dict, State]:
+    """Processes feedback from user and updates state with the feedback."""
+    result = {"feedback": feedback}
+    return result, state.update(**result)
+
+
+@action(reads=["current_draft", "feedback"], writes=["final_draft"])
+def final_result(state: State) -> Tuple[dict, State]:
+    """Returns the final result of the process."""
+    result = {"final_draft": state["current_draft"]}
+    return result, state.update(final_draft=result["final_draft"])
+
+
+def application(
+    app_id: str = None, project: str = "demo:email_assistant", username: str = None
+) -> Application:
+    builder = (
+        ApplicationBuilder()
+        .with_actions(
+            process_input,
+            determine_clarifications,
+            clarify_instructions,
+            formulate_draft,
+            process_feedback,
+            final_result,
+        )
+        .with_transitions(
+            ("process_input", "determine_clarifications"),
```

Comment:
these could also just be function pointers instead of strings.

Reply:
Yeah but its a bit easy to shoot yourself in the foot, as it depends on whether we're using the named or not-named version (E.G. `*args` from the above or `**kwargs`, or functions or action-objects).


Might be some syntactic sugar helps but we can't do the same thing we did with Hamilton as we're doing purely procedural (rather than declarative) graph-building.